File: lab1/report/lab1_listing.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <string>

std::vector<int> DEMAND = {0, 7, 6, 3, 0, 2};
#define INITIAL_NUMBER_OF_MACHINES 1
#define WAREHOUSE_CAPACITY 20
#define MAX_ORDER 5
#define DELIVERY_COST 50
#define PER_MACHINE_COST 10
#define STORAGE_COST 10

bool USE_MEMOIZATION = true;

int min_cost(
    int month,
    int warehouse,
    const std::vector<int>& demand,
    std::vector<std::vector<int>>& dp,
    std::vector<std::vector<bool>>& used,
    bool use_memoization = true,
    int final_warehouse = -1,
    int storage_cost = STORAGE_COST,
    int delivery_cost = DELIVERY_COST) {

    if (month == demand.size()) {
        if (final_warehouse == -1 || warehouse == final_warehouse) {
            return 0;
        }
        return INT_MAX;
    }

    if (use_memoization && used[month][warehouse]) {
        return dp[month][warehouse];
    }

    int best = INT_MAX;

    for (int order = 0; order <= MAX_ORDER; order++) {
        if (warehouse + order < demand[month]) {
            continue;
        }

        int next_warehouse = warehouse + order - demand[month];
        if (next_warehouse > WAREHOUSE_CAPACITY) {
            continue;
        }

        int step_cost = 0;
        if (order > 0) {
            step_cost += delivery_cost + PER_MACHINE_COST * order;
        }
        step_cost += storage_cost * next_warehouse;

        int future_cost = min_cost(
            month + 1,
            next_warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost
        );
        if (future_cost == INT_MAX) {
            continue;
        }

        int total = step_cost + future_cost;
        if (total < best) {
            best = total;
        }
    }

    if (use_memoization) {
        used[month][warehouse] = true;
        dp[month][warehouse] = best;
    }
    return best;
}

int machines_num(
    int month,
    int warehouse,
    const std::vector<int>& demand,
    std::vector<std::vector<int>>& dp,
    std::vector<std::vector<bool>>& used,
    bool use_memoization = true,
    int final_warehouse = -1,
    int storage_cost = STORAGE_COST,
    int delivery_cost = DELIVERY_COST) {

    if (month == demand.size()) {
        return 0;
    }

    int best = INT_MAX;
    int best_order = 0;

    for (int order = 0; order <= MAX_ORDER; order++) {
        if (warehouse + order < demand[month]) {
            continue;
        }

        int next_warehouse = warehouse + order - demand[month];
        if (next_warehouse > WAREHOUSE_CAPACITY) {
            continue;
        }

        int step_cost = 0;
        if (order > 0) {
            step_cost += delivery_cost + PER_MACHINE_COST * order;
        }
        step_cost += storage_cost * next_warehouse;

        int future_cost = min_cost(
            month + 1,
            next_warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost
        );
        if (future_cost == INT_MAX) {
            continue;
        }

        int total = step_cost + future_cost;
        if (total < best) {
            best = total;
            best_order = order;
        }
    }

    return best_order;
}

void minimize_cost(
    std::vector<int>& plan,
    int& total_cost,
    const std::vector<int>& demand,
    bool use_memoization = true,
    int final_warehouse = -1,
    int storage_cost = STORAGE_COST,
    int delivery_cost = DELIVERY_COST) {
    int warehouse = INITIAL_NUMBER_OF_MACHINES;
    std::vector<std::vector<int>> dp(
        demand.size() + 1,
        std::vector<int>(WAREHOUSE_CAPACITY + 1, 0)
    );
    std::vector<std::vector<bool>> used(
        demand.size() + 1,
        std::vector<bool>(WAREHOUSE_CAPACITY + 1, false)
    );

    total_cost = min_cost(
        0,
        warehouse,
        demand,
        dp,
        used,
        use_memoization,
        final_warehouse,
        storage_cost,
        delivery_cost
    );
    plan.clear();

    for (int month = 0; month < demand.size(); month++) {
        int order = machines_num(
            month,
            warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost
        );
        plan.push_back(order);
        warehouse = warehouse + order - demand[month];
    }
}

std::vector<int> task1(int final_warehouse = -1) {
    std::vector<int> plan;
    int total_cost = 0;

    minimize_cost(plan, total_cost, DEMAND, USE_MEMOIZATION, final_warehouse);

    std::cout << "Optimal order plan "
              << (final_warehouse == -1 ? "without constraint" : "with constraint")
              << " on final warehouse stock";

    if (final_warehouse != -1) {
        std::cout << " (" << final_warehouse << ")";
    }

    std::cout << ": ";

    for (int order : plan) {
        std::cout << order << " ";
    }
    std::cout << "\nTotal cost: " << total_cost << std::endl;

    return plan;
}

int task2(std::vector<int> base_plan = {}) {
    int base_cost = 0;
    if (base_plan.empty()) {
        minimize_cost(base_plan, base_cost, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST);
    }

    int left = STORAGE_COST;
    int right = STORAGE_COST;

    for (int cost = STORAGE_COST - 1; cost >= 0; cost--) {
        std::vector<int> current_plan;
        int current_cost = 0;
        minimize_cost(current_plan, current_cost, DEMAND, USE_MEMOIZATION, -1, cost);

        if (current_plan == base_plan) {
            left = cost;
        } else {
            break;
        }
    }

    for (int cost = STORAGE_COST + 1; ; cost++) {
        std::vector<int> current_plan;
        int current_cost = 0;
        minimize_cost(current_plan, current_cost, DEMAND, USE_MEMOIZATION, -1, cost);

        if (current_plan == base_plan) {
            right = cost;
        } else {
            break;
        }
    }

    std::cout << "Storage cost range: ["
              << left << ", " << right << "]" << std::endl;

    return 0;
}

int task3(std::vector<int> base_plan = {}) {
    int base_cost = 0;
    if (base_plan.empty()) {
        minimize_cost(base_plan, base_cost, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST, DELIVERY_COST);
    }

    int left = DELIVERY_COST;
    int right = DELIVERY_COST;

    int total_demand = 0;
    for (int value : DEMAND) {
        total_demand += value;
    }

    int required = total_demand - INITIAL_NUMBER_OF_MACHINES;
    int min_trips = (required + MAX_ORDER - 1) / MAX_ORDER;

    int base_trips = 0;
    for (int order : base_plan) {
        if (order > 0) {
            base_trips++;
        }
    }

    for (int cost = DELIVERY_COST - 1; cost >= 0; cost--) {
        std::vector<int> current_plan;
        int current_cost = 0;
        minimize_cost(current_plan, current_cost, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST, cost);

        if (current_plan == base_plan) {
            left = cost;
        } else {
            break;
        }
    }

    for (int cost = DELIVERY_COST + 1; ; cost++) {
        if (base_trips == min_trips) {
            std::cout << "No upper bound: the base plan uses "
                      << min_trips
                      << " trips, which is the minimum possible number of trips."
                      << std::endl;
        }
        right = -1;
        break;

        std::vector<int> current_plan;
        int current_cost = 0;
        minimize_cost(current_plan, current_cost, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST, cost);

        if (current_plan == base_plan) {
            right = cost;
        } else {
            break;
        }
    }

    std::cout << "Trip fixed cost range: ["
              << left << ", " << right << "]" << std::endl;

    return 0;
}

void task4(int final_warehouse) {
    auto plan = task1(final_warehouse);
    std::vector<int> remains = {};
    for (int i = 0; i < plan.size(); ++i) {
        remains.push_back(plan[i] - DEMAND[i]);
    }
    std::cout << "Stock at the beginning of each month: ";
    for (int remain : remains) {
        std::cout << remain << " ";
    }
    std::cout << std::endl;
}

int main(int argc, char* argv[]) {
    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--memo") {
            USE_MEMOIZATION = true;
        } else if (arg == "--no-memo") {
            USE_MEMOIZATION = false;
        } else {
            std::cout << "Unknown argument: " << arg << std::endl;
            std::cout << "Usage: ./lab1 [--memo | --no-memo]" << std::endl;
            return 1;
        }
    }

    std::cout << "\n================ Task 1 ================\n";
    auto plan = task1();

    std::cout << "\n================ Task 2 ================\n";
    task2(plan);

    std::cout << "\n================ Task 3 ================\n";
    task3(plan);

    std::cout << "\n================ Task 4 ================\n";
    task4(2);
    return 0;
}

File: lab1/report/lab1_listing.py
from __future__ import annotations

import sys

DEMAND = [0, 7, 6, 3, 0, 2]
INITIAL_NUMBER_OF_MACHINES = 1
WAREHOUSE_CAPACITY = 20
MAX_ORDER = 5
DELIVERY_COST = 50
PER_MACHINE_COST = 10
STORAGE_COST = 10

USE_MEMOIZATION = True


def min_cost(
    month: int,
    warehouse: int,
    demand: list[int],
    dp: list[list[int]],
    used: list[list[bool]],
    use_memoization: bool = True,
    final_warehouse: int = -1,
    storage_cost: int = STORAGE_COST,
    delivery_cost: int = DELIVERY_COST,
) -> int:
    if month == len(demand):
        if final_warehouse == -1 or warehouse == final_warehouse:
            return 0
        return sys.maxsize

    if use_memoization and used[month][warehouse]:
        return dp[month][warehouse]

    best = sys.maxsize

    for order in range(MAX_ORDER + 1):
        if warehouse + order < demand[month]:
            continue

        next_warehouse = warehouse + order - demand[month]
        if next_warehouse > WAREHOUSE_CAPACITY:
            continue

        step_cost = 0
        if order > 0:
            step_cost += delivery_cost + PER_MACHINE_COST * order
        step_cost += storage_cost * next_warehouse

        future_cost = min_cost(
            month + 1,
            next_warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost,
        )
        if future_cost == sys.maxsize:
            continue

        total = step_cost + future_cost
        if total < best:
            best = total

    if use_memoization:
        used[month][warehouse] = True
        dp[month][warehouse] = best

    return best


def machines_num(
    month: int,
    warehouse: int,
    demand: list[int],
    dp: list[list[int]],
    used: list[list[bool]],
    use_memoization: bool = True,
    final_warehouse: int = -1,
    storage_cost: int = STORAGE_COST,
    delivery_cost: int = DELIVERY_COST,
) -> int:
    if month == len(demand):
        return 0

    best = sys.maxsize
    best_order = 0

    for order in range(MAX_ORDER + 1):
        if warehouse + order < demand[month]:
            continue

        next_warehouse = warehouse + order - demand[month]
        if next_warehouse > WAREHOUSE_CAPACITY:
            continue

        step_cost = 0
        if order > 0:
            step_cost += delivery_cost + PER_MACHINE_COST * order
        step_cost += storage_cost * next_warehouse

        future_cost = min_cost(
            month + 1,
            next_warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost,
        )
        if future_cost == sys.maxsize:
            continue

        total = step_cost + future_cost
        if total < best:
            best = total
            best_order = order

    return best_order


def minimize_cost(
    plan: list[int],
    demand: list[int],
    use_memoization: bool = True,
    final_warehouse: int = -1,
    storage_cost: int = STORAGE_COST,
    delivery_cost: int = DELIVERY_COST,
) -> int:
    warehouse = INITIAL_NUMBER_OF_MACHINES
    dp = [[0] * (WAREHOUSE_CAPACITY + 1) for _ in range(len(demand) + 1)]
    used = [[False] * (WAREHOUSE_CAPACITY + 1) for _ in range(len(demand) + 1)]

    total_cost = min_cost(
        0,
        warehouse,
        demand,
        dp,
        used,
        use_memoization,
        final_warehouse,
        storage_cost,
        delivery_cost,
    )
    plan.clear()

    for month in range(len(demand)):
        order = machines_num(
            month,
            warehouse,
            demand,
            dp,
            used,
            use_memoization,
            final_warehouse,
            storage_cost,
            delivery_cost,
        )
        plan.append(order)
        warehouse = warehouse + order - demand[month]

    return total_cost


def task1(final_warehouse: int = -1) -> list[int]:
    plan: list[int] = []
    total_cost = minimize_cost(plan, DEMAND, USE_MEMOIZATION, final_warehouse)

    print(
        "Optimal order plan "
        + ("without constraint" if final_warehouse == -1 else "with constraint")
        + " on final warehouse stock",
        end="",
    )

    if final_warehouse != -1:
        print(f" ({final_warehouse})", end="")

    print(": ", end="")

    for order in plan:
        print(order, end=" ")
    print(f"\nTotal cost: {total_cost}")

    return plan


def task2(base_plan: list[int] | None = None) -> int:
    if base_plan is None:
        base_plan = []

    if not base_plan:
        minimize_cost(base_plan, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST)

    left = STORAGE_COST
    right = STORAGE_COST

    for cost in range(STORAGE_COST - 1, -1, -1):
        current_plan: list[int] = []
        minimize_cost(current_plan, DEMAND, USE_MEMOIZATION, -1, cost)

        if current_plan == base_plan:
            left = cost
        else:
            break

    cost = STORAGE_COST + 1
    while True:
        current_plan = []
        minimize_cost(current_plan, DEMAND, USE_MEMOIZATION, -1, cost)

        if current_plan == base_plan:
            right = cost
            cost += 1
        else:
            break

    print(f"Storage cost range: [{left}, {right}]")
    return 0


def task3(base_plan: list[int] | None = None) -> int:
    if base_plan is None:
        base_plan = []

    if not base_plan:
        minimize_cost(
            base_plan,
            DEMAND,
            USE_MEMOIZATION,
            -1,
            STORAGE_COST,
            DELIVERY_COST,
        )

    left = DELIVERY_COST
    right = DELIVERY_COST

    total_demand = 0
    for value in DEMAND:
        total_demand += value

    required = total_demand - INITIAL_NUMBER_OF_MACHINES
    min_trips = (required + MAX_ORDER - 1) // MAX_ORDER

    base_trips = 0
    for order in base_plan:
        if order > 0:
            base_trips += 1

    for cost in range(DELIVERY_COST - 1, -1, -1):
        current_plan: list[int] = []
        minimize_cost(current_plan, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST, cost)

        if current_plan == base_plan:
            left = cost
        else:
            break

    cost = DELIVERY_COST + 1
    while True:
        if base_trips == min_trips:
            print(
                "No upper bound: the base plan uses "
                f"{min_trips} trips, which is the minimum possible number of trips."
            )
        right = -1
        break

        current_plan = []
        minimize_cost(current_plan, DEMAND, USE_MEMOIZATION, -1, STORAGE_COST, cost)

        if current_plan == base_plan:
            right = cost
            cost += 1
        else:
            break

    print(f"Trip fixed cost range: [{left}, {right}]")
    return 0


def task4(final_warehouse: int) -> None:
    plan = task1(final_warehouse)
    remains: list[int] = []
    for i in range(len(plan)):
        remains.append(plan[i] - DEMAND[i])

    print("Stock at the beginning of each month: ", end="")
    for remain in remains:
        print(remain, end=" ")
    print()


def main() -> int:
    global USE_MEMOIZATION

    for arg in sys.argv[1:]:
        if arg == "--memo":
            USE_MEMOIZATION = True
        elif arg == "--no-memo":
            USE_MEMOIZATION = False
        else:
            print(f"Unknown argument: {arg}")
            print("Usage: ./lab1.py [--memo | --no-memo]")
            return 1

    print("\n================ Task 1 ================")
    plan = task1()

    print("\n================ Task 2 ================")
    task2(plan)

    print("\n================ Task 3 ================")
    task3(plan)

    print("\n================ Task 4 ================")
    task4(2)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
